### PR TITLE
Unit help - add a 'hasImage' to avoid unnecessary image rendering

### PR DESCRIPTION
**Pull request by DanVanAtta**
There are four parts to this update:
1. Simplify UnitStatsTable
2. Introduce and use "ImageKey" as a parameter object in UnitImageFactory
3. Replace the key for UnitImageFactory image cache with "ImageKey" instead of a string.
4. Add a 'hasImage' method and invoke that in UnitStatsTable. This avoids rendering images where we only need to check if they exist (and we do not use the resulting image).

## Functional Changes

[] New map or map update
[] New Feature
[] Feature update or enhancement
[] Feature Removal
[] Code Cleanup or refactor
[] Configuration Change
[x] Problem fix
[] Other:   

## Testing
- Did some smoke testing verify that both map loaded and unit help rendered on maps:
  - NWO
  - WaW
  - WofW

## Additional Notes to Reviewer

This series of updates seeks to address part of: https://github.com/triplea-game/triplea/issues/7142 (Calculating Data on "Unit Help" takes extraordinarily long time).

After this update I was able to load & view the "unit help" on WofW for the very first time ever (and not hit an out of heap error).

## Release Note

FIX|Reduce memory usage when loading the "unit help" menu

### CODE REVIEW

File: game-core/src/main/java/games/strategy/triplea/image/UnitImageFactory.java [hunk 1]
```diff
@@ -61,6 +65,113 @@ public UnitImageFactory(
...
+          .damaged(Matches.unitHasTakenSomeBombingUnitDamage().test(unit))
+          .disabled(Matches.unitIsDisabled().test(unit))
+          .build();
+    }
+
+    public String getFullName() {
+      return getBaseImageName() + player.getName();
+    }
+
+    public String getBaseImageName() {
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/image/UnitImageFactory.java:102]: Codacy found an issue: [The method 'getBaseImageName()' has an NPath complexity of 102606, current threshold is 200](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5910459)

**Comment by DanVanAtta** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 2
           

Coverage increased per file
===========================
+ game-core/src/main/java/games/strategy/engine/data/UnitType.java  76
         

Coverage decreased per file
===========================
- game-core/src/main/java/games/strategy/triplea/image/UnitImageFactory.java  -2
         

Complexity increasing per file
==============================
- game-core/src/main/java/games/strategy/triplea/ui/menubar/help/UnitStatsTable.java  1
- game-core/src/main/java/games/strategy/triplea/image/MissingImageException.java  1
         

Complexity decreasing per file
==============================
+ game-core/src/main/java/games/strategy/engine/data/UnitType.java  -14
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5910459&bid=19361187)

File: game-core/src/main/java/games/strategy/triplea/image/MissingImageException.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package games.strategy.triplea.image;
+
+import games.strategy.triplea.image.UnitImageFactory.ImageKey;
+import games.strategy.triplea.util.UnitCategory;
+
+public class MissingImageException extends RuntimeException {
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/image/MissingImageException.java:6]: Codacy found an issue: [Classes implementing Serializable should set a serialVersionUID](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5910459)

File: game-core/src/main/java/games/strategy/triplea/image/UnitImageFactory.java (hunk 1 above)
Comment (DanVanAtta) [reply game-core/src/main/java/games/strategy/triplea/image/UnitImageFactory.java:102]: NPath complexity of 102K!!  Does this take the cake?  :cake:

**Comment by DanVanAtta** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 1
           

Complexity increasing per file
==============================
- game-core/src/main/java/games/strategy/triplea/ui/menubar/help/UnitStatsTable.java  1
- game-core/src/main/java/games/strategy/triplea/image/MissingImageException.java  1
         

Complexity decreasing per file
==============================
+ game-core/src/main/java/games/strategy/engine/data/UnitType.java  -14
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5910459&bid=19361187)

File: game-core/src/main/java/games/strategy/triplea/image/UnitImageFactory.java (hunk 1 above)
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/triplea/image/UnitImageFactory.java:102]: Codacy found an issue: [The method 'getBaseImageName()' has an NPath complexity of 102606, current threshold is 200](https://app.codacy.com/gh/triplea-game/triplea/pullRequest?prid=5910459)

File: game-core/src/main/java/games/strategy/triplea/ui/menubar/help/UnitStatsTable.java [hunk 3]
```diff
@@ -103,15 +103,13 @@ public static String getUnitStatsTable(final GameData gameData, final UiContext
     }
     final Set<UnitType> all = data.getUnitTypeList().getAllUnitTypes();
     all.removeAll(unitsSoFar);
+    unitTypes.put(
+        GamePlayer.NULL_PLAYERID,
+        getPlayerUnitsWithImages(GamePlayer.NULL_PLAYERID, data, uiContext));
+    unitsSoFar.addAll(unitTypes.get(GamePlayer.NULL_PLAYERID));
+    all.removeAll(unitsSoFar);
     if (!all.isEmpty()) {
```
Comment (RoiEXLab) [inline game-core/src/main/java/games/strategy/triplea/ui/menubar/help/UnitStatsTable.java:109]: Could you explain why this if statement is always true?
If all is no subset of unitsSoFar this isn't always the case

**Comment by DanVanAtta** [comment]
Look to an earlier commit. We had an always true param that was or'd with
the not empty check. I removed just the param and not the if statement
(that was incorrect, the whole if was a tautology since we had  `true ||`)

**Comment by RoiEXLab** [comment]
@DanVanAtta Not seeing it. Are you sure we're talking abou the same piece of code?

**Comment by DanVanAtta** [comment]
@RoiEXLab https://github.com/triplea-game/triplea/pull/7249/commits/5c1718f385c61b64198f6ea3c36deae1c764d99f#diff-e4f55d25cc44f84a9875c84fd14f408dL111

**Comment by RoiEXLab** [comment]
Ah, thanks for clarifying 👍

**Pull request merged**